File: pages/login_page.py
from playwright.sync_api import Page

from pages.base_page import BasePage


class LoginPage(BasePage):
    URL = "/parabank/index.htm"

    USERNAME_INPUT = 'input[name="username"]'
    PASSWORD_INPUT = 'input[name="password"]'
    LOGIN_BUTTON = 'input[value="Log In"]'
    ERROR_MESSAGE = "p.error"
    REGISTER_LINK = 'a[href*="register.htm"]'

    def __init__(self, page: Page, base_url: str) -> None:
        super().__init__(page, base_url)

    def open(self) -> "LoginPage":
        self.navigate(self.URL)
        return self

    def login(self, username: str, password: str) -> None:
        self.fill(self.USERNAME_INPUT, username, "username")
        self.fill(self.PASSWORD_INPUT, password, "password")
        self.click(self.LOGIN_BUTTON, "Log In button")

    def is_on_login_page(self) -> bool:
        return "index.htm" in self.page.url or self.page.url.endswith("/parabank/")

    def is_logged_in(self) -> bool:
        return "overview.htm" in self.page.url

    def has_error(self) -> bool:
        return self.page.locator(self.ERROR_MESSAGE).count() > 0

    def is_username_field_visible(self) -> bool:
        return self.page.locator(self.USERNAME_INPUT).is_visible()

File: dotnet/ParabankQa.Tests/Pages/LoginPage.cs
using Allure.Net.Commons;
using Microsoft.Playwright;

namespace ParabankQa.Tests.Pages;

public sealed class LoginPage(IPage page, string baseUrl) : BasePage(page, baseUrl)
{
    public const string Url = "/parabank/index.htm";

    private const string UsernameInput = "input[name=\"username\"]";
    private const string PasswordInput = "input[name=\"password\"]";
    private const string LoginButton = "input[value=\"Log In\"]";
    private const string ErrorMessage = "p.error";
    public const string RegisterLink = "a[href*=\"register.htm\"]";

    public async Task<LoginPage> OpenAsync()
    {
        await NavigateAsync(Url);
        return this;
    }

    public Task LoginAsync(string username, string password) =>
        AllureApi.Step($"Log in as '{username}'", async () =>
        {
            await FillAsync(UsernameInput, username, "username");
            await FillAsync(PasswordInput, password, "password");
            await ClickAsync(LoginButton, "Log In button");
        });

    public bool IsOnLoginPage() =>
        Page.Url.Contains("index.htm") || Page.Url.TrimEnd('/').EndsWith("/parabank");

    public bool IsLoggedIn() => Page.Url.Contains("overview.htm");

    public Task<bool> HasErrorAsync() =>
        Page.Locator(ErrorMessage).CountAsync().ContinueWith(t => t.Result > 0);

    public Task<bool> IsUsernameFieldVisibleAsync() => Page.Locator(UsernameInput).IsVisibleAsync();
}
